DB_HW_MountaineeringClub

skalantaryan opened the pull request.

skalantaryan commented:
> > In github it doesn't allow me to make changes to the files.
> 
> Honestly, I don't understand what kind of your issue is. You should have remote repo in github, local repo on your laptop. You make changes locally, commit and push them to remote. Please check these: https://learn.epam.com/catalog/detailsPage?id=601f195a-d408-4439-a16d-0630ed2a412e https://learn.epam.com/catalog/detailsPage?id=16d7f2e7-cc80-4870-928e-400723f732bb

Now I understand.
Thank you very much for your reply.
Next time I will do that correctly))

krabts reviewed (changes requested):
Hi Sona,

thank you for the HW provided. 

1. `Climbers`
     - please add UNIQUE constraint for phone_number into Doc description, it should be the same as on your ERD
     - could you please clarify why did you add climber_id to `Addresses` table but not address_id to `Climbers`? 
     - why did you choose 1:1 relation type? is it possible that two brothers have one address for example?
        ![image](https://github.com/user-attachments/assets/1d0f5c39-7171-43de-8029-85fbf33acb15)

2. `Addresses`
     - please add building number as a separate column, full address is not 1NF 

3. `Clubs`
     - let's imagine that two clubs will have the same name by accident. they should be identified by unique phone or email. If the name is unique, please also add such constraint

4. `Club_Climber`
     - I didn't find information about PK or any constraints for this table
     - can one climber be presented twice in the same club?

5. `Climb_Participants`
     - could you please clarify how PK from `Climbs` could be a newly created PK for this table?
        ![image](https://github.com/user-attachments/assets/b8f1d539-0486-4ffd-bcc4-d06e3c4162e4)

6. `Mountains`
     - country attribute is to be presented as FK from `country` table
     - please create new entity for difficulty, it's a good practice for different types, categories, etc.

7. `Equipment`
     - please present type as FK to separate entity
     - check your description
       ![image](https://github.com/user-attachments/assets/041060b4-5425-4cc4-a6f7-eb631ddd9535)
     - please consider CHECK constrains for price_per_day (it should be > 0, I guess) and quantity_available (>= 0, but not nullable from my point of view)
     - if you have quantity_available in `equipment`, maybe it's nice to have some rental quantity here? or we can imagine that a lot of the same equipment pieces have different IDs an we don't need to count their availability in both tables (we can check start-end dates). hope, you understand me

8. `Climb_Equipment`
     - well, quantity_used attribute depends on you choice that I described above.

9. `Payments`
     - payment method is to be separated as an entity
     - it's discussable, but I would prefer to have booking_id in `Payments` but not paiment_id in `Bookings`. Booking it's a independent entity, Payment is to be done for some reason ans booking is the reason for that.

I have some other remarks, but let's start from these. Don't worry, check my comments step by step, everything is fine ;)
If you have questions or objections, please let me know. 

Regards,
Kiryl

skalantaryan commented:
> Hi Sona,
> 
> thank you for the HW provided.
> 
> 1. `Climbers`
>    
>    * please add UNIQUE constraint for phone_number into Doc description, it should be the same as on your ERD
>    * could you please clarify why did you add climber_id to `Addresses` table but not address_id to `Climbers`?
>    * why did you choose 1:1 relation type? is it possible that two brothers have one address for example?
>      ![image](https://private-user-images.githubusercontent.com/157591026/424441338-1d0f5c39-7171-43de-8029-85fbf33acb15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzODcyODMsIm5iZiI6MTc0MjM4Njk4MywicGF0aCI6Ii8xNTc1OTEwMjYvNDI0NDQxMzM4LTFkMGY1YzM5LTcxNzEtNDNkZS04MDI5LTg1ZmJmMzNhY2IxNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOVQxMjIzMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT03Mzk3ZmRjMzhlOTAzZjcxNzRiN2MxYjM1YTVjZjU1MTQwYTIyZWZiMWU5MmEyOWNlNjRmMGVmZmMxZGEwN2EyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-nFefGmwwX8BVd7KHOKHwvH_aB8Zg8JEPymorzdesJ0)
> 2. `Addresses`
>    
>    * please add building number as a separate column, full address is not 1NF
> 3. `Clubs`
>    
>    * let's imagine that two clubs will have the same name by accident. they should be identified by unique phone or email. If the name is unique, please also add such constraint
> 4. `Club_Climber`
>    
>    * I didn't find information about PK or any constraints for this table
>    * can one climber be presented twice in the same club?
> 5. `Climb_Participants`
>    
>    * could you please clarify how PK from `Climbs` could be a newly created PK for this table?
>      ![image](https://private-user-images.githubusercontent.com/157591026/424448482-b8f1d539-0486-4ffd-bcc4-d06e3c4162e4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzODcyODMsIm5iZiI6MTc0MjM4Njk4MywicGF0aCI6Ii8xNTc1OTEwMjYvNDI0NDQ4NDgyLWI4ZjFkNTM5LTA0ODYtNGZmZC1iY2M0LWQwNmUzYzQxNjJlNC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOVQxMjIzMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lYTk0N2Y2YzU5MWU3YzA3ZjhjNDEwMzVmZjQ5ZDNhYTVkMjNiOTFhM2JkMGRkMjg4ZGQ1NDI5NzI5NDc0ZjM5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.6YW9m7omXz2k9xiYqAiebdQ7EBbHmhkLkzukkI8NKF4)
> 6. `Mountains`
>    
>    * country attribute is to be presented as FK from `country` table
>    * please create new entity for difficulty, it's a good practice for different types, categories, etc.
> 7. `Equipment`
>    
>    * please present type as FK to separate entity
>    * check your description
>      ![image](https://private-user-images.githubusercontent.com/157591026/424454750-041060b4-5425-4cc4-a6f7-eb631ddd9535.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIzODcyODMsIm5iZiI6MTc0MjM4Njk4MywicGF0aCI6Ii8xNTc1OTEwMjYvNDI0NDU0NzUwLTA0MTA2MGI0LTU0MjUtNGNjNC1hNmY3LWViNjMxZGRkOTUzNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMzE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDMxOVQxMjIzMDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wNGQwYmJiYjg2MDgxZWFiM2MzMDA1ZjYyMzU2NjcxMjdiN2MwZTI2ODkxYWZiMTE4Y2M3NTE4ZTE1MzEyMTFlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.lPNV7SLtJzzAlMdNlSYZFaSdEC39d2Rj6k1QzF2CCqE)
>    * please consider CHECK constrains for price_per_day (it should be > 0, I guess) and quantity_available (>= 0, but not nullable from my point of view)
>    * if you have quantity_available in `equipment`, maybe it's nice to have some rental quantity here? or we can imagine that a lot of the same equipment pieces have different IDs an we don't need to count their availability in both tables (we can check start-end dates). hope, you understand me
> 8. `Climb_Equipment`
>    
>    * well, quantity_used attribute depends on you choice that I described above.
> 9. `Payments`
>    
>    * payment method is to be separated as an entity
>    * it's discussable, but I would prefer to have booking_id in `Payments` but not paiment_id in `Bookings`. Booking it's a independent entity, Payment is to be done for some reason ans booking is the reason for that.
> 
> I have some other remarks, but let's start from these. Don't worry, check my comments step by step, everything is fine ;) If you have questions or objections, please let me know.
> 
> Regards, Kiryl

Dear Kiryl,
Thank you for your review.
1. About the Climbers table:
- I had added address_id to table Climbers and climber_id to Address table and now I see that there is not need of that. So I will remove the climber_id column from Address table.
3. About the Clubs table:
- I made the club_name column unique.
4. About the Club_Climber table:
- I made club_id and climber_id composite primary keys. That way each climber can be in a specific club only once (no duplicate lines).
- Each climber can be a member of multiple clubs.
- Each club can have multiple members.
5. About the Climb_Participants table:
- The serial data type was a mistake. I was not attentive enough. I made it Int.
6. About the Mountains table:
- Drawsql says I have reached my limit of 15 tables. So I will try to create another diagram using another tool. But I have updated the docx file and added the Difficulty table there.
7. About the Equipment table:
- Again about the type entity I have added it on docx file but not in the ERD diagram. 
- I didn't understand the rental_qunatity part quite well. In the equipment_rental table I have columns referring to rental_date and return_date, if that's what you mean. Should I also add rental_quantity here?
9 About the Payments table:
- I have added the Payment_methods table in docx file.
- As you mentioned I removed payment_id from Bookings table and added booking_id to Payments table. I agree with you that this way it is more logical.

I will update the docx file in a minute.
I hope this time it will be better))

Best regards,
Sona Kalantaryan

krabts commented:
Hi Sona,

thank you for the updates made. I have several comments.

1. `Mountains`
    - it has M:M relationship with `country` on your ERD, does one mountain is located in different countries?
    - even if yes, M:M is to be implemented using bridge table

2. `Climb_Participants`
    - please check your composite PK in doc, ERD looks fine

3. `Equipment_Rental`
    - ok, just add rental_quantity here to align with other equipment related tables

I need to check you HW one more time more attentively, but you can make these changes if you don't have questions or objections.

Regards,
Kiryl

skalantaryan commented:
> Hi Sona,
> 
> thank you for the updates made. I have several comments.
> 
> 1. `Mountains`
>    
>    * it has M:M relationship with `country` on your ERD, does one mountain is located in different countries?
>    * even if yes, M:M is to be implemented using bridge table
> 2. `Climb_Participants`
>    
>    * please check your composite PK in doc, ERD looks fine
> 3. `Equipment_Rental`
>    
>    * ok, just add rental_quantity here to align with other equipment related tables
> 
> I need to check you HW one more time more attentively, but you can make these changes if you don't have questions or objections.
> 
> Regards, Kiryl

krabts commented:
> Good day, thank you for your review.
> 
> As a mountain can be located in the border of 2 countries and can be climbed from both countries, I made the relationship between Mountains and Country M:M. About the equipment_rental_quantity: I have rental_fee in equipment_rental table and I have price_per_day in equipment table. Thus I thought maybe it would be better to replace the rental_fee with equipment_rental_quantity, as the rental_fee is calculated by this equation: equipment_rental_quantity * price_per_day. In this case the equation will be used for payments table if the payment refers to equipment rental and it will take price_per_day from equipment table and equipment-rental_quantity from equipment_rental table and calculate amount. So what do you think is better in this case? Does my solution satisfy 3NF?
> 
> Best regards, Sona Kalantaryan

Well, if I got you correctly, your suggestion is the following?

`equipment.price_per_day` * diff (`equipment_rental.rental_date` - `equipment_rental.return_date`) * `equipment_rental.quantity` = `payments.amount`

What are your doubts about 3NF?

skalantaryan commented:
> > Good day, thank you for your review.
> > As a mountain can be located in the border of 2 countries and can be climbed from both countries, I made the relationship between Mountains and Country M:M. About the equipment_rental_quantity: I have rental_fee in equipment_rental table and I have price_per_day in equipment table. Thus I thought maybe it would be better to replace the rental_fee with equipment_rental_quantity, as the rental_fee is calculated by this equation: equipment_rental_quantity * price_per_day. In this case the equation will be used for payments table if the payment refers to equipment rental and it will take price_per_day from equipment table and equipment-rental_quantity from equipment_rental table and calculate amount. So what do you think is better in this case? Does my solution satisfy 3NF?
> > Best regards, Sona Kalantaryan
> 
> Well, if I got you correctly, your suggestion is the following?
> 
> `equipment.price_per_day` * diff (`equipment_rental.rental_date` - `equipment_rental.return_date`) * `equipment_rental.quantity` = `payments.amount`
> 
> What are your doubts about 3NF?

Yes that's what I was suggesting.
So is this ok?
Should I modify my ERD this way?

krabts commented:
Well, in plane view it looks fine. In general, every piece of equipment should have it's inventory number not to calculate it's quantity manually, like in `dvd_rental` database. But for your solution it's enough.

skalantaryan commented:
> Well, in plane view it looks fine. In general, every piece of equipment should have it's inventory number not to calculate it's quantity manually, like in `dvd_rental` database. But for your solution it's enough.

Ok, thanks.

krabts commented:
Hi Sona, hope you are doing well. I am waiting for your `DML. SELECT Statement` HW :)

krabts commented:
My bad, deadline is tomorrow. Good luck!

skalantaryan commented:
I will upload it tomorrow I think. Lately I have been a little busy. I managed to work only on the diagram.

krabts commented:
Ok, it's up to you. Just keep me updated.

skalantaryan commented:
Should I now merge this pull request?

krabts commented:
> Should I now merge this pull request?

No, I will check your HW and let you know.

krabts reviewed (approved):
Hi Sona,

thank you for the updates made. We will be able make changes during the DDL task if needed. Now you can merge this.

Regards,
Kiryl
